V1

ryansmith3136 opened the pull request:
Lpxc started off as a reference implementation for Heroku Add-on providers to use as an example. However, there have been several companies (including Heroku) who have started to use this code in production. Thus, I have been motivated to add the following features to the library in hopes of making it a suitable choice for production systems.
- Tests - There are a couple unit tests and an integration test.
- Multi-tenancy - It is now possible to use 1 instance of **lpxc** to transport logs for many logplex tokens.

I am also planning on merging in the changes from #3 into the V1 release.

cc: @deafbybeheading @technomancy

Commit:
add support for multiple tenants

Commit:
add integration test

msakrejda commented on lpxc.rb:
```diff
@@ -6,40 +6,59 @@
+    @msgid = opts[:msgid] || "-"
+
+    #This will show up in the Heroku logs tail command as: app[lpxc]
+    @procid = opts[:procid] || "lpxc"
```
Nice.

technomancy replied:
Yeah, this is better the way in #3 which just takes procid as an arg to `#start`.

msakrejda commented on lpxc.rb:
```diff
@@ -6,40 +6,59 @@
+
+    #In most cases, the value should be: https://east.logplex.io/logs
+    err = "Must set logplex url."
+    @logplex_url = URI(opts[:logplex_url] || ENV["LOGPLEX_URL"] || raise(err))
```
Also nice.

msakrejda commented on lpxc.rb:
```diff
@@ -6,40 +6,59 @@
   #Messages are flushed from memory every 500ms or when we have 300 messages,
   #whichever comes first.
-  def self.start
+  def start
```
Is `#start` necessary? Given that there is no matching stop, or any other semantics for controlling the worker threads, can't this just happen implicitly on first `puts`? Is that crazy?

msakrejda commented on lpxc.rb:
```diff
@@ -48,61 +67,75 @@ def self.start
               req.add_field('Content-Type', 'application/logplex-1')
               resp = conn.request(req)
-              $stdout.puts("at=request-sent status=#{resp.code}")
+              $stdout.puts("at=req-sent status=#{resp.code}") if ENV['DEBUG']
```
Nice.

msakrejda commented on readme.md:
```diff
@@ -1,19 +1,17 @@
-Lpxc.Start
-Lpxc.puts(logplex_token, "hello world")
+lpxc = Lpxc.new
+lpxc.puts("hello world", 't.123')
```
Unless you decide to make `#start` implicit as per the above suggestion, it should be restored here.

msakrejda commented:
Excellent. Despite the relative rawness, `lpxc` has been working pretty well for us so far. I'm very excited by having a more mature structure to the project.

technomancy commented on lpxc.rb:
```diff
@@ -48,61 +67,75 @@ def self.start
+  def fmt(data)
     pkt = "<190>1 "
-    pkt += "#{data[:ts]} "
+    pkt += "#{data[:t].to_datetime.rfc3339.to_s} "
```
The `to_datetime` method is new in 1.9, so you'll want to use the `strftime` invocation from #3 here instead. That and the literal hash syntax in `puts` were the only things preventing bamboo compatibility IIRC.

Commit:
use a time layout that is ruby < 1.9 friendly

Commit:
make ruby 1.8.7, 1.9.3, and 2.0.0 compatible

ryansmith3136 commented:
@technomancy Thanks for the pointers on 1.8.7 compatibility. I implemented your suggestions and the test are now passing on that version of ruby.

ryansmith3136 commented:
Before we cut the first version of this code, I would like to address the start/stop issue. I am not sure I know the best way to handle this. I am going to play around with a few different approaches and propose a suggestion later today.

Commit:
dry up test server

Commit:
Introduce batch sizing.

You can specify a batch size when initializing a client.
When the number of log messages buffered equals the batch size,
the client will submit them to the server.

Commit:
correct api border and make all stdout/stderr output optional

Commit:
improve flush behaviour and tests

Commit:
make a note about reqs_in_flight

Commit:
make a note about wait

Commit:
improve docs

Commit:
add gemspec

Commit:
link to docs

ryansmith3136 commented:
Thanks for the feedback guys. I continued to make improvements and cut a release.
